Overlays: Add info boxes

kristofbolyai opened the pull request:
Supersed of #279.
![image](https://user-images.githubusercontent.com/49001742/192536871-840f2870-c3ee-411a-80f3-358053d82554.png)
![image](https://user-images.githubusercontent.com/49001742/192536909-23de32f5-8017-4f45-a59e-be2dc145ff9c.png)
![image](https://user-images.githubusercontent.com/49001742/192537027-26d758f0-0dd8-41af-bcca-e82b75a2b7d6.png)

kristofbolyai commented on src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java:
```diff
@@ -86,6 +86,12 @@
+    @Setting(displayName = "Show Guild Manage Member Search Bar", description = "Should the search bar be shown in the guild manage members GUI?\n\n§8Members that match the search will be highlighted.", order = 24)
+    public boolean showGuildMemberManageSearchbar = true;
+
+    @Setting(displayName = "Searched Item Highlight Colour", description = "What colour should the highlight for searched items be?\n\n§aClick the coloured box to open the colour wheel.", order = 25)
```
Incorrect text here

DonkeyBlaster replied:
Wait where?

kristofbolyai replied:
It says searched item highlight color and not guild menu highlight color

DonkeyBlaster replied:
The one for guild territory says the same; I'll change them both

kristofbolyai replied:
Also perhaps use this color in territory manage overlay aswell if it doesnt have a separate setting

CringeAura commented on src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java:
```diff
@@ -86,6 +86,12 @@
     @Setting(displayName = "Searched Item Highlight Colour", description = "What colour should the highlight for searched items be?\n\n§aClick the coloured box to open the colour wheel.", order = 23)
     public CustomColor guildTerritoryMenuSearchHighlightColor = new CustomColor(80, 242, 242);
 
+    @Setting(displayName = "Show Guild Manage Member Search Bar", description = "Should the search bar be shown in the guild manage members GUI?\n\n§8Members that match the search will be highlighted.", order = 24)
```
Do you know if the name of the setting is getting truncated in the menu? If it is, it should be shortened.

DonkeyBlaster replied:
Yup a few of these are getting cut - fixed in next commit

HighCrit commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/GuildMemberManageOverlay.java:
```diff
@@ -0,0 +1,126 @@
+        searchField.mouseClicked(offsetMouseX, offsetMouseY, e.getMouseButton());
+
+        // Only listen for left clicks
+        if (e.getMouseButton() != 0) {
+            return;
+        }
```
Should this be higher in the function?

ryanzhoudev committed:
initial (buggy) implementation of colors & boxes

Commit:
chore: spotless formatting

ryanzhoudev committed:
Use existing methods, improve & color code check

Didn't realize existing methods could still take §codes. Switch to those, add \ check to cancel color parse

ryanzhoudev committed:
don't add unnecessary method

ryanzhoudev committed:
Merge branch 'main' into info-boxes

ryanzhoudev committed:
Allow formatting codes

ie bold, italics, etc

ryanzhoudev committed:
Merge branch 'main' into info-boxes

Commit:
chore: spotless formatting

ryanzhoudev committed:
Merge branch 'main' into info-boxes

ryanzhoudev committed:
Merge branch 'main' into info-boxes

ryanzhoudev committed:
fix import

ryanzhoudev committed:
theoretically working info vars

Commit:
chore: spotless formatting

ryanzhoudev committed:
Merge branch 'main' into info-boxes

kristofbolyai committed:
Merge branch 'main' into info-boxes

# Conflicts:
#	common/src/main/java/com/wynntils/core/features/FeatureRegistry.java
#	common/src/main/resources/assets/wynntils/lang/en_us.json

kristofbolyai committed:
InfoBoxesFeature: Finish

magicus commented on common/src/main/java/com/wynntils/functions/MinecraftFunctions.java:
```diff
@@ -12,28 +12,28 @@ public class MinecraftFunctions {
         @Override
         public Integer getValue(String argument) {
-            return McUtils.player().getBlockX();
+            return McUtils.player() == null ? null : McUtils.player().getBlockX();
```
How can McUtils.player() be null? I thought it was set as soon as you logged on to the server.

kristofbolyai replied:
For some reason, this was called on feature init. As I am writing this, because this is what the render preview used to be. And it used to be a static field..., so this was needed to be calculated. I fixed that so this is not needed anymore technically. I am not sure if there is value to keeping this?

kristofbolyai replied:
(Also, functions do not fail gracefully :))

magicus replied:
I thought I actually wrote a rationale about skipping functions: They are so simple I did not expect them to fail, and if they do, the failure will propagate to the `/function` command and/or InfoBoxOverlay, which will fail gracefully.

But it might be a point in disabling individual functions. If one function is behaving bad, we do not have to bring down the entire InfoBoxOverlay. I'll fix that.

kristofbolyai replied:
This case was more of bad use of functions than them being bad, but I think there is still value to disabling functions.

magicus replied:
And no, if it's not needed we should not keep it. I've been trying to kill code that does null checks for player() elsewhere. If we have such code, people start thinking it's needed and add it in some places but not other. We must be able to trust this value, period.

Also, functions should perhaps be blocked by the FunctionManager if we are not connected to a wynncraft server, just like events are not posted. That way we can definitely not do anything wrong with this.

kristofbolyai replied:
Reverted this

magicus commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -87,7 +87,13 @@
   "feature.wynntils.healthPotionBlocker.threshold.name": "Threshold (%%)",
   "feature.wynntils.healthPotionBlocker.thresholdReached": "You are already at %s%% of your max health!",
-  "feature.wynntils.hidePotionGlint.name": "Hide Potion Glint",
+  "feature.wynntils.hidePotionGlint.name": "Hide Potion Glint Feature",
```
This seem like a merge error, old code reappearing.

magicus commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -87,7 +87,13 @@
+  "feature.wynntils.infoBoxes.overlay.infoBox.content.description": "What should be displayed in the info box?",
+  "feature.wynntils.infoBoxes.overlay.infoBox.content.name": "Info Box Content",
+  "feature.wynntils.infoBoxes.overlay.infoBox.name": "Info Box %d",
+  "feature.wynntils.infoBoxes.overlay.infoBox.textShadow.description": "Should the overlay for teleport scrolls be drawn with a shadow?",
```
Teleport scrolls FTW! 😆

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+import net.minecraft.client.resources.language.I18n;
+
+@FeatureInfo(category = FeatureCategory.OVERLAYS)
+public class InfoBoxesFeature extends UserFeature {
```
Just a bit on the naming. I understand that `InfoBoxesFeature` is probably more correct than `InfoBoxFeature`, but the latter sounds much better. And I don't think we've had plural for other features either, like "BossBars"?

kristofbolyai committed:
I18n: Fix merge

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+public class InfoBoxesFeature extends UserFeature {
+
+    @OverlayInfo(renderType = RenderEvent.ElementType.GUI)
+    private final Overlay infoBox1Overlay = new InfoBoxOverlay(1);
```
Oh, it's still hard-coded. :( I'd hoped you would have made this more general. Oh well, that's for another time, I guess.

kristofbolyai replied:
This would require GUI work, and overlay systematical changes.

kristofbolyai committed:
I18n: Fix text

kristofbolyai committed:
Rename class

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+            functionDependencies.clear();
+
+            StringBuilder builder = new StringBuilder(renderableText.length() + 10);
+            Matcher m = INFO_VARIABLE_PATTERN.matcher(renderableText);
```
There is a `FunctionManager.doFormat()` which uses the new `{function}` syntax. But I see now that I never finished `getStringFromTemplate()`, which is that you would have needed here.

I am not entirely sure I want to import the old `%info` syntax. We can possibly do this short-term if you really want to push this now, as long as we agree that this is just temporary.

kristofbolyai replied:
I did not even know that was a thing. This is mostly code I took from DonkeyBlaster, and made it work. Yes, I remember all the new things you want to do with function variables, but most of that code right now is TODO. So, until that is changed, we should push this.

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+
+                m.appendReplacement(builder, replacement);
+            }
+            m.appendTail(builder);
```
In any case, if you do keep this functionality, it should move to FunctionManager, to something like `getStringFromLegacyTemplate()`...

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+                            .withTextShadow(textShadow));
+        }
+
+        private String parseColorCodes(String toProcess) {
```
This too, if kept, should move out of the Feature itself. Either to FunctionManager (if we think this is strictly something we will do for function templates), or to like StringUtils or ComponentUtils (which include support for coded strings as well)

magicus commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxesFeature.java:
```diff
@@ -0,0 +1,218 @@
+
+            // FIXME: We do re-calculate this on render, but this is preview only, and fixing this would need a lot of
+            // architectural changes at the moment
+            TextRenderTask toRenderPreview = recalculateFunctions("&cX: %x%, &9Y: %y%, &aZ: %z%");
```
I don't get this. How does the normal toRender get updated? Only on config changes? Even if it does not need recalculating on every rendered frame, it do need some periodic update.. Or am I missing something?

kristofbolyai replied:
Wait..., I need to do this in toRender too otherwise we are getting old data. Well..

magicus replied:
May I suggest instead doing the calculation in a separate function that listens to ClientTickEvent, and cache it as a pre-calculated string ever so often?

And also perhaps expose a Config that says how often values should be recalculated, either in terms of ticks or expressed in seconds (the latter is perhaps easier for users, but requires you to convert it into ticks). I think a good default value is about 4 times per second, i.e. after 5 ticks.

magicus replied:
Most functions are relatively cheap, but some iterate through all inventory parsing lore, and that's no good to make for each frame we're drawing.

kristofbolyai committed:
Review changes

kristofbolyai committed:
MinecraftFunctions: Remove null-checks

kristofbolyai committed:
InfoBoxFeature: Use caching

kristofbolyai committed:
Merge branch 'main' into info-boxes

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/overlays/InfoBoxFeature.java:
```diff
@@ -0,0 +1,159 @@
+        public void render(PoseStack poseStack, float partialTicks, Window window) {
+            if (!WynnUtils.onWorld()) return;
+
+            if (System.nanoTime() - lastUpdate > secondsPerRecalculation * 1e+9) {
```
@magicus I hope you do not mind this. Overlays are not registered to the event bus... This is a good enough replacement IMO.

magicus replied:
Yes, that is perfectly fine, perhaps even better than my suggestion. (Otherwise, if we really need events, let the owning Feature listen for them. I've done so in one or two places).

magicus replied:
Maybe do the same to preview?

kristofbolyai replied:
Yeah, but ticking 6 different overlays is pretty messy from a feature event.

kristofbolyai replied:
I think it is fine for previews to be updated every time, for now. If there comes some really heavy variable, we can update it. (also, the active function by nature uses some caching to prevent per render call calculation)

magicus replied:
Ok, fine. (We should perhaps also let the preview use hard-coded functions from us, since if the user has weird strings it might now show up properly. But that's a future improvement, if we should do it at all.)

kristofbolyai replied:
> Ok, fine. (We should perhaps also let the preview use hard-coded functions from us, since if the user has weird strings it might now show up properly. But that's a future improvement, if we should do it at all.)

It does have a hard coded value?

magicus replied:
Oh, I missed that.
